[GTK][WPE][Skia] Fix threading issue with DisplayList rendering / GraphicsContextState merging

nikolaszimmermann opened the pull request:
#### b4d253a7cf351a3e84efed18f2dc53d557195971
<pre>
[GTK][WPE][Skia] Fix threading issue with DisplayList rendering / GraphicsContextState merging
<a href="https://bugs.webkit.org/show_bug.cgi?id=282084">https://bugs.webkit.org/show_bug.cgi?id=282084</a>

Reviewed by Said Abou-Hallawa.

Pattern is not ThreadSafeRefCounted - therefore using patterns in a
multithreaded environment as we do for Skia multithreaded CPU rendering
is unsafe, leading to assertions in ASSERT_ENABLED builds - fix that by
making Pattern inherit from ThreadSafeRefCounted instead of RefCounted.

Covered by existing tests.

* Source/WebCore/platform/graphics/Pattern.h:

Canonical link: <a href="https://commits.webkit.org/286744@main">https://commits.webkit.org/286744@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52fa32a47899dee13fde2b5cd57c46997190724f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76909 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55944 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29815 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81452 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28185 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79026 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65086 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4237 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60265 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18346 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79976 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50213 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66020 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40580 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47615 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23515 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26511 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68733 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23846 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82889 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4285 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2858 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68542 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4439 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65993 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67796 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16911 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11780 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9864 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4232 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7045 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4255 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7686 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6013 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa reviewed (approved):
I do not think this PR is covered by existing tests. This multi-threaded painting is new to WebKit. Do you plan to have an EWS bot for Skia?

nikolaszimmermann committed:
[GTK][WPE][Skia] Fix threading issue with DisplayList rendering / GraphicsContextState merging
https://bugs.webkit.org/show_bug.cgi?id=282084

Reviewed by Said Abou-Hallawa.

Pattern is not ThreadSafeRefCounted - therefore using patterns in a
multithreaded environment as we do for Skia multithreaded CPU rendering
is unsafe, leading to assertions in ASSERT_ENABLED builds - fix that by
making Pattern inherit from ThreadSafeRefCounted instead of RefCounted.

Covered by existing tests.

* Source/WebCore/platform/graphics/Pattern.h:

Canonical link: https://commits.webkit.org/286744@main
